ansible-galaxy - enforce string versions and fix installing prereleases from scm sources

s-hertel opened the pull request:
##### SUMMARY

Fix unhandled tracebacks in the dep resolver by enforcing string versions when loading Requirements.

Fix installing prereleases from virtual dependencies without `--pre` by adding them as direct requests before resolving dependencies.

Fixes #79109
Fixes #79168
Fixes #78067

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
ansible-galaxy

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -442,7 +442,7 @@ def is_satisfied_by(self, requirement, candidate):
         # NOTE: and one of them is a pre-release that also matches a
         # NOTE: transitive dependency of another requirement.
-        allow_pre_release = self._with_pre_releases or not (
+        allow_pre_release = self._with_pre_releases or requirement.is_virtual or not (
```
This is probably not harmful, but I'm pretty sure it does not address the root cause of the bug described in https://github.com/ansible/ansible/issues/78067 and https://github.com/ansible/ansible/issues/79109.

The root cause of this bug is insufficient input validation. The whole resolution logic is built around the version being a string. When it is not, the broken assumptions start causing problems. I'm pretty sure, a similar bug may happen in other places that treat the version as a string.

The proper resolution could be conversion of any version inputs to strings. This should happen outside the dependency resolver so that the resolver could keep relying on this expectation in any place where this is happening. Inputs should be validated/transformed to a canonical state in an outer abstraction layer of a system so that the internal/core modules could rely on those inputs being of a predictable type/form.

It looks like this bug is coming from this line https://github.com/ansible/ansible/blob/cb2e434/lib/ansible/galaxy/dependency_resolution/dataclasses.py#L285 — if a version key is missing, it'll be `'*'` (as in "any version"). That's a string. But if it is present, there is no validation for the actual value.
It should probably be changed to also turn `null` and empty string versions to "any" asterisk `*`:
```diff
-       req_version = collection_req.get('version', '*')
+       input_version_value = collection_req.get('version')
+       req_version = str(
+           input_version_value  # int
+           if isinstance(input_version_value, int)
+           else (input_version_value or '*')  # `None` or `''` == `*`
+       )
```

It would be a good idea to have an assertion that it's a string in the data structure initializer, at https://github.com/ansible/ansible/blob/cb2e434/lib/ansible/galaxy/dependency_resolution/dataclasses.py#L558.

As a possibility of a future refactoring, it could be nice to have a data structure/class/abstraction representing a version, instead of a plain string. It'd be much easier to reason about such a dedicated data type if it's used all across the resolver code base.

s-hertel replied:
I must have only scanned https://github.com/ansible/ansible/issues/78067, because I didn't realize it was `type: dir`. You're right, it doesn't fix https://github.com/ansible/ansible/issues/78067, it only impacts virtual collections, ie git and subdirs. `subdirs` shouldn't have any version associated other than '*', so this could be tightened up a bit. If it is `type: subdirs` or `type: dir`, an integer version should probably be an error as suggested in the issue.

This does fix https://github.com/ansible/ansible/issues/79109. This prerelease logic is to determine if 1) --pre flag was provided or 2) the version is an exact match. In the case of virtual collections, 2) should always be true. Below the prerelease logic, virtual collections are handled correctly without version comparison, and True is returned.

I think comparing a stringified virtual version would introduce a few issues.

Mainly, if it's not an artifact version, comparing it like it is one creates all sorts of weird cornercases. If someone has a branch name containing `<` or even looks like a semver version because it contains `.`, that should have absolutely no bearing on prereleases - it's not a semver version even if it looks like it. It's unrelated to the artifact version that will be built and installed.

I'm not sure from_requirement_dict is the correct place to do the casting. It will break a bunch of valid-version tests (there are tests for null, float, list versions, but clearly not comprehensive for `type: git`). Maybe simplest is to give an error if a `type: git` version is not a string in from_requirement_dict.

I think we may just need to exclude scm candidate/requirements from normal version evaluation (which is attempted in most places) and use == to determine if they match.

> It should probably be changed to also turn null and empty string versions to "any" asterisk *:

I don't see the need for another 'any' placeholder, but it does seem like the '*' default is problematic for `type: git`.

It should probably be HEAD instead. What about scm requirements that have '*' explicitly though in their requirements.yml? (I'm wondering now how the dependency resolver handles incompatible scm requirements - guessing it doesn't properly). Should these resolve successfully?
```yaml
- name: file://path/to/repo
  version: tagname
  type: git
- name: file://path/to/repo
  version: "*"
  type: git
```
I'd expect these to incompatible I think:
```yaml
- name: file://path/to/repo
  version: tagname
  type: git
- name: file://path/to/repo
  version: HEAD
  type: git
```

There is a somewhat related issue where the real version is replaced with '*' here https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/concrete_artifact_manager.py#L316 and it's 
 helping cause #79168. The dep resolver doesn't consider virtual collections and their dependencies 'user-requested' when they are.

Much more complicated than I thought at a glance, guessing there's some invasive re-working of the scm stuff to be done here. :exploding_head: Thanks for being a sounding board, really appreciate the review.

webknjaz replied:
These are interesting observations. It does indeed sound like this needs a refactoring with better test coverage.

s-hertel replied:
After thinking more about this, I think the simple solution is just
* enforce str versions for scm requirements earlier
* add virtual dependencies to direct requirements to fix the pre-release logic

Leaving this for another PR, but since I was wondering about comparing virtual versions above, I'm thinking it might make more sense to compare `type: dir` using `verify_local_collection`ish logic. If `find_matches` did the same when evaluating preinstalled candidates, virtual collections could support `--force` too instead of always reinstalling.

webknjaz commented on test/integration/targets/ansible-galaxy-collection-scm/tasks/multi_collection_repo_individual.yml:
```diff
@@ -1,5 +1,14 @@
+    content: |
+      collections:
+        - name: file://{{ test_repo_path }}/.git#collection_2
+          type: git
+          version: 12345
```
Could you add a test with an integer version for a non-virtual collection? I wonder if it'll cause a similar traceback.

s-hertel replied:
There are some existing tests for non-virtual invalid versions in test/integration/targets/ansible-galaxy-collection-scm/tasks/test_invalid_version.yml. It's in the scm test suite, but it's the virtual's dependency (i.e. `type: dir`) with the invalid version.

s-hertel replied:
I was initially fixing the unhandled traceback here (by avoiding the cause), but I've changed it to be an error instead and moved the test case into the file with all the other invalid version tests - scm versions need to be quoted.

s-hertel committed:
load virtual dependencies to handle as direct requests before running the dep resolver

s-hertel committed:
more version validation

s-hertel committed:
remove extraneous change

changelog

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -504,8 +504,20 @@ def build_collection(u_collection_path, u_output_path, force):
     return collection_output
 
 
+def process_virtual_requirements(requirement, resolved_requirements, artifacts_manager):
```
This name is confusing to me:
1) It seems to only process one virtual requirement at a time (so why use plural?)
2) Saying “process” is too generic, it doesn't really suggest what the implication of such processing is?

Looking at the implementation, it seems like it attempts to extract non-virtual dependencies. Maybe we could rename it?

Although, I don't quite like that it mutates an incoming param instead of returning or yielding stuff. I'd say that yielding would be more suiting here.

s-hertel replied:
It's plural because it's finding all leaf nodes for an ephemeral collection. So it should end up with possibly multiple `type: dir` requirements.

And yeah, didn't think too long about the naming. If you have a preference, feel free to suggest a new name.

s-hertel replied:
Now it's in providers and called `get_virtual_requirement_dependencies`, and I fixed it to be a generator instead of mutating an incoming param.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -504,8 +504,20 @@ def build_collection(u_collection_path, u_output_path, force):
+    # type: (Requirement, t.Iterable[Requirement], ConcreteArtifactsManager) -> None
+    requirement.validate_version()
+
+    if requirement.is_virtual:
```
Refactoring opportunity: invert the condition and dedent a block of code.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -504,8 +504,20 @@ def build_collection(u_collection_path, u_output_path, force):
+    requirement.validate_version()
+
+    if requirement.is_virtual:
+        for dep_name, dep_req in artifacts_manager.get_direct_collection_dependencies(requirement).items():
+            dep = Requirement.from_requirement_dict({'name': dep_name, 'version': dep_req}, artifacts_manager)
+            process_virtual_requirements(dep, resolved_requirements, artifacts_manager)
```
I've got a bad feeling about the recursion here and a dependency resolution attempt outside the dependency resolver... Pretty sure the resolution should be scoped to the resolver.

s-hertel replied:
Okay, I actually went back and forth about that. My first draft was contained in the resolver, but I was modifying the preferred candidates during resolution, which seemed like it was going to be problematic. We calculate the preferred candidates outside of the dep resolver, but perhaps I can move that inside as a first step...

s-hertel replied:
I've refactored so it occurs when the dependency resolver is instantiated and put a safeguard against recursive dependencies. Since things external and internal to the dependency resolver needed this info, and `_resolve_depenency_map` is the only place that has access to the provider, it's now responsible for 1. passing the parsed requirements back into `resolve()`
and 2. handling the unsatisfied logic.

I'd like to rewrite the unsatisfied logic, but I will leave that for a follow up PR that won't be backported.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -522,6 +534,14 @@ def download_collections(
     :param no_deps: Ignore any collection dependencies and only download the base requirements.
     :param allow_pre_release: Do not ignore pre-release versions when selecting the latest.
     """
+    for install_req in list(collections):
```
This doesn't need to be converted to a list if it's already expected to be a list according to the typings.

s-hertel replied:
I'm mutating the list, don't want it changing while iterating, so I copied it.

webknjaz replied:
In that case, consider making the intent explicit:
```suggestion
    for install_req in collections.copy():
```

webknjaz replied:
Or
```suggestion
    for install_req in collections[:]:
```
_(whichever syntax you prefer)_

webknjaz replied:
This is because using `list()` suggests that the purpose is data conversion rather than copying, and therefore it is confusing.

s-hertel replied:
:+1: Removed mutating the list while refactoring.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -551,6 +557,20 @@ def is_online_index_pointer(self):
     def source_info(self):
         return self._source_info
 
+    def validate_version(self):
```
Maybe call this
```suggestion
    def assert_version(self):
```
since it's going to be raising exceptions where it's used?

s-hertel replied:
Renamed it to `assert_version`.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -551,6 +557,20 @@ def is_online_index_pointer(self):
+            version_req = f"A string version is required. Got {self.ver} ({type(self.ver)})."
+        version_err = f"Invalid version found for the collection '{self}'. {version_req}"
+        if not isinstance(self.ver, string_types):
+            raise AnsibleError(version_err)
```
Can we use specific exceptions for things raised in the internals and only convert them into `AnsibleError` on the outer layer?
A suitable exception type for this case could be a `TypeError`, for example.

s-hertel replied:
Okay. This was just moving the existing exception, but I can look at raising a TypeError instead.

s-hertel replied:
Fixed.

s-hertel committed:
Move finding ephemeral requirement dependencies and handling --force into the dep resolver's __init__

Give the same message regardless of codepath when all collections are already installed and update test.

Reword error message for malformed collection versions so it's consistent regardless of codepath

Improve handling pre-releases for already installed collections and pre-release requirements

Update changelog

s-hertel committed:
Remove opinion on irregularity to keep things backwards compatible

s-hertel committed:
update unit tests

fix sanity tests

s-hertel committed:
remove extra import

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -674,60 +674,10 @@ def install_collections(
-        # NOTE: No need to include signatures if the collection is already installed
         Candidate(coll.fqcn, coll.ver, coll.src, coll.type, None)
-        for coll in preferred_requirements
+        for coll in find_existing_collections(output_path, artifacts_manager)
```
We shouldn't start scanning the disk `if force_deps` — all this is being discarded on the next line in 50% of cases.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -428,11 +430,13 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
             req_source, req_type,
             req_signature_sources,
         )
+        req.assert_version()
```
Making assertions when an object holding invalid data is already initialized is an anti-pattern. This should be checked in or before init...

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -551,6 +555,28 @@ def is_online_index_pointer(self):
+        name = self.fqcn if self.fqcn is not None else self
+        version_err = f"Invalid version found for the collection '{name}': {self.ver} ({type(self.ver)}). {version_req}"
+        if not isinstance(self.ver, string_types):
+            raise TypeError(version_err)
```
How about a less invasive approach? https://github.com/ansible/ansible/pull/81694
